src/main/java/com/aws/greengrass/clientdevices/auth/iot/registry/ThingRegistry.java

```diff
@@ -5,33 +5,94 @@
 
 package com.aws.greengrass.clientdevices.auth.iot.registry;
 
+import com.aws.greengrass.clientdevices.auth.api.DomainEvents;
 import com.aws.greengrass.clientdevices.auth.exception.CloudServiceInteractionException;
 import com.aws.greengrass.clientdevices.auth.iot.Certificate;
 import com.aws.greengrass.clientdevices.auth.iot.IotAuthClient;
 import com.aws.greengrass.clientdevices.auth.iot.Thing;
+import com.aws.greengrass.clientdevices.auth.iot.events.ThingUpdated;
 
 import java.util.Collections;
 import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import javax.inject.Inject;
 
 public class ThingRegistry {
-    // holds mapping of thingName to IoT Certificate ID;
+    // holds mapping of thingName to IoT Certificate IDs;
     // size-bound by default cache size, evicts oldest written entry if the max size is reached
-    private final Map<String, String> registry = Collections.synchronizedMap(
-            new LinkedHashMap<String, String>(RegistryConfig.REGISTRY_CACHE_SIZE, 0.75f, false) {
+    private final Map<String, Thing> registry = Collections.synchronizedMap(
+            new LinkedHashMap<String, Thing>(RegistryConfig.REGISTRY_CACHE_SIZE, 0.75f, false) {
                 @Override
                 protected boolean removeEldestEntry(Map.Entry eldest) {
                     return size() > RegistryConfig.REGISTRY_CACHE_SIZE;
                 }
             });
 
     private final IotAuthClient iotAuthClient;
+    private final DomainEvents domainEvents;
 
     @Inject
-    public ThingRegistry(IotAuthClient iotAuthClient) {
+    public ThingRegistry(IotAuthClient iotAuthClient, DomainEvents domainEvents) {
         this.iotAuthClient = iotAuthClient;
+        this.domainEvents = domainEvents;
+    }
+
+    /**
+     * Create a Thing.
+     * @param thingName ThingName
+     * @return Thing object
+     */
+    public Thing createThing(String thingName) {
+        return updateThing(Thing.of(thingName));
+    }
+
+    /**
+     * Retrieve a Thing based on ThingName.
+     *
+     * @param thingName ThingName
+     * @return Thing domain object, if it exists
+     */
+    public Thing getThing(String thingName) {
+        return Thing.of(getThingInternal(thingName));
+    }
+
+    /**
+     * Update a Thing.
+     * @param thing Thing which is being updated
+     * @return      New Thing version
+     */
+    public Thing updateThing(Thing thing) {
+        // TODO - need to throw exception if provided Thing is not the most recent version
+        Thing oldThing = getThingInternal(thing.getThingName());
+
+        if (oldThing == null) {
+            storeThing(thing);
```

Comment:
I think here we would also want to increment the version. Every time we store a thing or make a change in the registry we return a new version.

Reply:
I actually removed this code -- I'm forcing callers to call `createThing` first. I agree with the versioning approach.